edb/edgeql/compiler/polyres.py

```diff
@@ -611,6 +655,241 @@ def _get_cast_distance(
+        isinstance(func, s_func.Function)
+        and func.get_server_param_conversions(schema)
+    ):
+        curr_server_param_conversions: dict[
+            str,
+            dict[str, tuple[irast.PathId, irast.Param, list[str]]],
+        ] = {}
+
+        arg_conversions: dict[str, str | list[str]] = json.loads(
+            arg_conversions_json
+        )
+        for arg_name, conversion_info in arg_conversions.items():
+            if isinstance(conversion_info, str):
+                conversion_name = conversion_info
+            else:
+                conversion_name = conversion_info[0]
+
+            # Get the arg being converted
+            arg_key: int | str
+            arg: tuple[s_types.Type, irast.Set]
+            param, arg_key, arg = _get_arg(
+                func_params,
+                arg_name,
+                args,
+                kwargs,
+                error_msg=f'Server param conversion {conversion_name} error',
+                schema=schema,
+            )
+
+            if arg[1].expr is None:
+                # Dummy set, do nothing
+                continue
+
+            original_type: s_types.Type = arg[0]
+            if original_type != param.get_type(schema):
+                # Wrong param type, function candidate doesn't apply.
+                # TODO: Check "any" params
+                return None
+
+            if not isinstance(arg[1].expr, irast.Parameter):
+                if original_type.is_array():
+                    # Array literals are extracted as expressions
+                    raise errors.QueryError(
+                        f"Argument '{arg_name}' must be query parameter",
+                        span=arg[1].expr.span,
+                    )
+                else:
+                    raise errors.QueryError(
+                        f"Argument '{arg_name}' "
+                        f"must be constant or query parameter",
+                        span=arg[1].expr.span,
+                    )
+
+            # Get info about the conversion
+            converted_type: s_types.Type
+            additional_info: list[str] = []
+            if conversion_name == 'cast_int64_to_str':
+                converted_type = schema.get(
+                    'std::str', type=s_scalars.ScalarType
+                )
+
+            elif conversion_name == 'cast_int64_to_float64':
+                converted_type = schema.get(
+                    'std::float64', type=s_scalars.ScalarType
+                )
+
+            elif conversion_name == 'join_str_array':
+                assert isinstance(conversion_info, list)
+                separator = conversion_info[1]
+
+                converted_type = schema.get(
+                    'std::str', type=s_scalars.ScalarType
+                )
+                additional_info = [separator]
+
+            else:
+                raise RuntimeError(
+                    f'Unknown server param conversion: {conversion_name}'
+                )
```

Comment:
Is there a way to make this more data driven? In terms of cleanness, I'd make a ranking like:
1. Declared in the schema somehow
2. Declared in a table in the code somehow
3. Done in its own function, at least

1 might not be worth it, but doing either 2 or 3 definitely is. We don't want this table of operations buried in the middle of a big and complex function.